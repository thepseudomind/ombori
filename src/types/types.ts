export interface IMain{
    status: boolean,
    toggleSidebar?: any
}

export interface IProduct{
    id: number,
    email: string,
    first_name: string,
    last_name: string,
    avatar: string
}

export type TLoader = {
    mini: boolean
}

export type TMessage = {
    message: string
}