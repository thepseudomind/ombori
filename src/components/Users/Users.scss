@import '../../mixins/mixins';

@keyframes rotate{
    0%{
        transform: rotate(0deg);
    }

    100%{
        transform: rotate(360deg);
    }
}

@keyframes fadeInOut{
    0%{
        transform: translateY(-2rem);
        opacity: 0;
        visibility: hidden;
    }

    50%{
        transform: translateY(0);
        opacity: 1;
        visibility: visible;
    }

    100%{
        transform: translateY(0);
        opacity: .5;
        visibility: hidden;
    }
}

.users{
    width: 100%;
    height: 90vh;
    padding: 2rem 0;
    position: relative;
    overflow-y: scroll;

    &::-webkit-scrollbar{
        display: none;
    }

    &__spinner{
        height: 3rem;
        width: 3rem;
        border: .2rem solid rgba(black, .2);
        border-left: .2rem solid black;
        border-radius: 50%;
        animation: rotate 1s linear infinite;
        position: fixed;
        bottom: 1rem;
        left: 2rem;
    }
}

.row{
    width: 70%;
    margin: 0 auto;
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;
    align-items: center;
    align-content: center;

    @include respond (tab-land){
        width: 80%;
        justify-content: space-around;
    }

    @include respond (phone){
        width: 70%;
    }

    &.active{
        justify-content: space-around;
    }
}