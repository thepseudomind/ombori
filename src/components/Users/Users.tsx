import React, {useState, useEffect} from 'react';
import './Users.scss';
import User from '../User/User';
import Loader from '../Loader/Loader';
import { IMain, IProduct } from '../../types/types';
import NotificationBox from '../NotificationBox/NotificationBox';

const Users = ({status} : IMain)=>{
    enum loadMoreUsers { inactive=0, loading=1, foundUsers=2, noMoreUsers = 3};
    const [totalUsers, setTotal] = useState(0);
    const [fetchedUsers, setUsers] = useState([]);
    const [usersToRender, addToRender] = useState([]);
    const [loadingMoreUsers, loadingStatus] = useState(loadMoreUsers[0]);
    const [pageToFetch, changePage] = useState(1);

    // Fetch first batch of users
    useEffect(()=>{
        setUsers([]);
        setTimeout(()=>{
            fetch(`https://reqres.in/api/users?page=${pageToFetch}`)
            .then(
                res => res.json()
            ).then(
                results => {
                    setUsers(results.data);
                    setTotal(results.total);
                    console.log(results.data);
                }
            );
        },3000);
    }, []);

   // Users to render
   useEffect(()=>{
        let productsCollected : [] = [];
        fetchedUsers.forEach((v, i)=>{
            productsCollected.push(v);
        });
        addToRender(productsCollected);
   }, [fetchedUsers]);
   
   // Add to the users to be rendered
    const addToProducts = (start : number)=>{
        try {
            loadingStatus(loadMoreUsers[1]);
            setTimeout(()=>{
                fetch(`https://reqres.in/api/users?page=${start+1}`)
                .then(
                    res => res.json()
                ).then(
                    results => {
                        if((results.data.length + start) < totalUsers){
                            let productsCollected : [] = [];
                            results.data.forEach((v: never, i : number)=>{
                                productsCollected.push(v);
                            });
                            addToRender([...usersToRender, ...productsCollected]);
                            changePage(start + 1);
                            loadingStatus(loadMoreUsers[0]);
                        }
                    }
                );
            }, 2000);  
        } catch (e) {
            throw e; 
        }
    } 

   // Detect when user has reached end of products 
   const handleScroll = (e: any)=>{
        if(e.target.offsetHeight + e.target.scrollTop === e.target.scrollHeight){
            (totalUsers === usersToRender.length) ? loadingStatus(loadMoreUsers[3]) : addToProducts(pageToFetch);
        }
        // Reseting loading status
        setTimeout(()=>{
            loadingStatus(loadMoreUsers[0]);
        },2000)
   }
   
   // Render preloader upon page load (no products) or fetched Product list when available
    return (
        <div className="users" onScroll={handleScroll}>
            <div className={`row${(status) ? ' active' : ''}`}>
                {
                    (fetchedUsers.length === 0) ? <Loader mini={false}/> :
                    usersToRender.map((v : IProduct, i : number) => <User key={v.id} id={v.id} email={v.email} first_name={v.first_name} last_name={v.last_name} avatar={v.avatar}/> )
                }
            </div>
            {loadingMoreUsers === loadMoreUsers[1] ? <Loader mini={true}/> : <span></span> /* Show mini preloader when fetching more users */} 
            {loadingMoreUsers === loadMoreUsers[3] ? <NotificationBox message={loadMoreUsers[3]}/> : <span></span> /* Show notification when there's no more users */}
        </div>
    );
}

export default Users;