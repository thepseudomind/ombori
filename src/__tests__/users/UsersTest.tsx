import React from 'react';
import { shallow, mount } from  'enzyme';
import Users from '../../components/Users/Users';
import { IMain } from '../../types/types';

describe('userlist tests', ()=>{
    it('rendering users', ()=>{
        const mockProps : IMain = {
            status: true
        }
        const wrapper = shallow(<Users {...mockProps}/>);
        expect(wrapper).toMatchSnapshot();
    });
});