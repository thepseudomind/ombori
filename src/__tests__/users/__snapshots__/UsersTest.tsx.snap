// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userlist tests rendering users 1`] = `
<div
  className="users"
  onScroll={[Function]}
>
  <div
    className="row active"
  >
    <Loader
      mini={false}
    />
  </div>
  <span />
  <span />
</div>
`;
